packages/hash/frontend/src/pages/_app.page.tsx

```diff
@@ -131,18 +129,110 @@ const App: FunctionComponent<AppProps> = ({
         };
       `}
       />
-    </ApolloProvider>
+    </>
   );
 };
 
-const AppWithTypeSystemContextProvider: FunctionComponent<AppProps> = (
+// Based on the definition of `NextPage`
+export type AppPage<P = {}, IP = P> = NextComponentType<AppContext, IP, P>;
+
+const AppWithTypeSystemContextProvider: AppPage<AppProps, AppInitialProps> = (
   props,
-) => (
-  <TypeSystemContextProvider>
-    <SessionProvider>
-      <App {...props} />
-    </SessionProvider>
-  </TypeSystemContextProvider>
-);
+) => {
+  const { initialAuthenticatedUser } = props;
+
+  return (
+    <TypeSystemContextProvider>
+      <ApolloProvider client={apolloClient}>
+        <AuthInfoProvider initialAuthenticatedUser={initialAuthenticatedUser}>
+          <SessionProvider>
+            <App {...props} />
+          </SessionProvider>
+        </AuthInfoProvider>
+      </ApolloProvider>
+    </TypeSystemContextProvider>
+  );
+};
+
+const redirectInGetInitialProps = (params: {
+  appContext: AppContext;
+  location: string;
+}) => {
+  const {
+    appContext: {
+      ctx: { req, res },
+      router,
+    },
+    location,
+  } = params;
+
+  if (req && res) {
+    res.writeHead(307, { Location: location });
+    res.end();
+  } else {
+    void router.push(location);
+  }
+};
+
+// The list of page pathnames that should be accessible whether or not the user is authenticated
+const publiclyAccessiblePagePathnames = [
+  "/[account-slug]/[page-slug]",
+  "/login",
+  "/signup",
+];
+
+AppWithTypeSystemContextProvider.getInitialProps = async (appContext) => {
```

Comment:
Note to reviewers that [it's not possible to use `getServerSideProps` in `_app`](https://nextjs.org/docs/advanced-features/custom-app#caveats)